### PR TITLE
Add integration tests to check core functionality

### PR DESCRIPTION
**Pull request by ocadaruma**
## Summary
- This PR adds integration tests to check processing guarantees which are expected Decaton to always satisfy
- By this tests, we can make sure that future code changes don't break processing semantics of Decaton
- [ProcessingGuarantee.java](https://github.com/line/decaton/pull/35/files#diff-76bb76bb3bdb59bf6dca0f19c07e0a04R33) contains the interface for guarantee and enums of available guarantees.
- [ProcessorTestSuite](https://github.com/line/decaton/pull/35/files#diff-75717a8b58a436b83a3fcecd1d91733dR57) is the class that performs integration test
- For how to use `ProcessorTestSuite`, please see [CoreFunctionalityTest](https://github.com/line/decaton/pull/35/files#diff-e7d47fe3e6d50565c8109a715d581436R31)

### CODE REVIEW

File: testing/src/main/java/com/linecorp/decaton/testing/processor/ReprocessAwareOrdering.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+                    }
+                    excludeReprocess.removeLast();
+                }
+            } else {
+                excludeReprocess.add(task);
+            }
+            currentOffset = offset;
+        }
+
+        assertTrue(produced.equals(new ArrayList<>(excludeReprocess)));
```
Comment (kawamuray) [inline testing/src/main/java/com/linecorp/decaton/testing/processor/ReprocessAwareOrdering.java:110]: why not just `assertEquals`?

File: testing/src/main/java/com/linecorp/decaton/testing/processor/ReprocessAwareOrdering.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    @Override
+    public void onProcess(ProcessedRecord record) {
+        processedRecords.add(record);
+    }
+
+    @Override
+    public void doAssert() {
+        Map<String, List<TestTask>> perKeyProducedRecords = new HashMap<>();
```
Comment (kawamuray) [inline testing/src/main/java/com/linecorp/decaton/testing/processor/ReprocessAwareOrdering.java:50]: why don't you just build those datastructures from the beginning at `onXX`?

File: testing/src/main/java/com/linecorp/decaton/testing/processor/TestTask.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+ * Intended to be assigned unique id for each tasks so that every task can be differentiated.
+ */
+@Value
+public class TestTask {
+    /**
+     * Unique id of the task
+     */
+    String id;
+
+    private static final ObjectMapper MAPPER = new ObjectMapper();
```
Comment (kawamuray) [inline testing/src/main/java/com/linecorp/decaton/testing/processor/TestTask.java:42]: I think in our convention those kind of  "final yet mutable" objects are still named lower case.

File: testing/src/main/java/com/linecorp/decaton/testing/processor/ProcessorTestSuite.java [hunk 4]
```diff
@@ -0,0 +1,268 @@
+package com.linecorp.decaton.testing.processor;
```
[the remark this answers is not in the document]
Comment (ocadaruma) [reply testing/src/main/java/com/linecorp/decaton/testing/processor/ProcessorTestSuite.java:1]: Oops why....

File: processor/src/it/java/com/linecorp/decaton/processor/CoreFunctionalityTest.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+                            Thread.currentThread().interrupt();
+                            throw new RuntimeException(e);
+                        } finally {
+                            completion.complete();
+                        }
+                    }, executorService);
+                }))
+                // If task is completed asynchronously there's no guarantee about
+                // completion ordering nor serial processing
+                .semantics(EnumSet.of(
```
[the remark this answers is not in the document]
Comment (ocadaruma) [reply processor/src/it/java/com/linecorp/decaton/processor/CoreFunctionalityTest.java:70]: Yeah I also considered about that, but I think there might be some semantics which is not included in default, but be enabled in "opt-in" manner (e.g. `at most once`)

To support such semantics, we'll add `includeSemantics`, and it's not so different from custom built-list ?

File: testing/src/main/java/com/linecorp/decaton/testing/processor/ProcessorTestSuite.java [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+        private final KafkaClusterRule rule;
+
+        /**
+         * Configure test-specific processing logic
+         */
+        private Function<ProcessorsBuilder<TestTask>, ProcessorsBuilder<TestTask>> configureProcessorsBuilder;
+        /**
+         * Configure retry-queueing feature for the subscription
+         */
+        private RetryConfig retryConfig;
```
[the remark this answers is not in the document]
Comment (ocadaruma) [reply testing/src/main/java/com/linecorp/decaton/testing/processor/ProcessorTestSuite.java:81]: This builder also allows to custom `ProcessorsBuilder` and `PropertySupplier`.

If we pass partially-built `SubscriptionBuilder`, users have to be carefully not to overwrite `processorsBuilder` because of following reasons. So it's not a good I/F I think?

- we're setting `preprocessor` as head of processor pipeline to record processing time
- supplied `processorsBuilder` has to consume `TestTask`

File: testing/src/main/java/com/linecorp/decaton/testing/processor/ReprocessAwareOrdering.java (hunk 1 above)
Comment (ocadaruma) [reply testing/src/main/java/com/linecorp/decaton/testing/processor/ReprocessAwareOrdering.java:110]: this reason
https://github.com/line/decaton/pull/18#discussion_r394786361

File: testing/src/main/java/com/linecorp/decaton/testing/processor/ReprocessAwareOrdering.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+public class ReprocessAwareOrdering implements ProcessingGuarantee {
```
[the remark this answers is not in the document]
Comment (ocadaruma) [reply testing/src/main/java/com/linecorp/decaton/testing/processor/ReprocessAwareOrdering.java:32]: Exactly.
> I think the intention of "Reprocess" here is that when fail-over occurs offset 1 could be re-processed by different instance which causes it to be processed after 2

I prefixed this guarantee with `ReprocessAware` because of following reason:
- If we support `at most once` semantics in the future, the ordering will be checked "strictly" (i.e. 1 never appears after 2)
- Then, we call such (strict) ordering as "ProcessOrdering" I think?

File: processor/src/it/java/com/linecorp/decaton/processor/CoreFunctionalityTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (kawamuray) [reply processor/src/it/java/com/linecorp/decaton/processor/CoreFunctionalityTest.java:70]: Hmm, I see.
As we talked before, I was thinking this "default it framework" as a shared function to test decaton's core/basic functionality that are considered to be applied for all instances (except few exceptions).
So that while
- no message loss
- ordering preserved
- serial processing
are reasonable to be applied for all the cases unless user configures it to opt-ted out trading it off with something else. On the other hand it sounds bit wired to me that this framework has a built-in check for at-most-once because it's a "feature" outside of core functionality that users configure with trade off with potential message loss.
So I feel like at-most-once check should be defined and manually performed in the class "AtMostOnceTest".
However, I just came up with that even for testing that we'll need some parts of this framework anyway - perform restart, rebalance to simulate some "basic events" that likely happen on typical deployment and some information collection as its interface already defined. 
So what about following approach?
- Still keep the current semantics list as required default
- Add a interface to allow users (it authors) inject custom impl of semantics check that is outside of framework built-in but still gets ran in the same way with built-in checks.

File: testing/src/main/java/com/linecorp/decaton/testing/processor/ProcessorTestSuite.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (kawamuray) [reply testing/src/main/java/com/linecorp/decaton/testing/processor/ProcessorTestSuite.java:81]: Hm, ok sound good so far then.
Actually I was thinking that the current CoreFunctionalityTest should also run the same test with configuring processors with 2 different scopes "partition" and "thread' respectively, and I thought its required to access SubscriptionBuilder directly but just turns out that its actually configuration of `ProcessorsBuilder`.

File: testing/src/main/java/com/linecorp/decaton/testing/processor/ReprocessAwareOrdering.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (kawamuray) [reply testing/src/main/java/com/linecorp/decaton/testing/processor/ReprocessAwareOrdering.java:32]: I see?  I was thinking that we'll just add a test checking exactly no duplicate in delivery for at-most-once case but apparently your test can also be added. Fine with as-is then. Please just add and enhance javadoc for it with much detailed explanation of its semantic.

File: testing/src/main/java/com/linecorp/decaton/testing/processor/ReprocessAwareOrdering.java (hunk 1 above)
Comment (kawamuray) [reply testing/src/main/java/com/linecorp/decaton/testing/processor/ReprocessAwareOrdering.java:110]: I see. then
- as a hint for assertion we would add on top `assertEquals` for their size?
- `//noinspection SimplifiableJUnitAssertion` please

File: processor/src/it/java/com/linecorp/decaton/processor/CoreFunctionalityTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (ocadaruma) [reply processor/src/it/java/com/linecorp/decaton/processor/CoreFunctionalityTest.java:70]: I see, makes sense.
> So I feel like at-most-once check should be defined and manually performed in the class "AtMostOnceTest"

Sounds nice.
> Add a interface to allow users (it authors) inject custom impl

like this?
```java
ProcessingGuarantee atMostOnce = new AtMostOnce();

ProcessorTestSuite.builder(rule)
                  .excludeSemantics(AT_LEAST_ONCE_DELIVERY)
                  .customSemantics(atMostOnce)
...
```

File: processor/src/it/java/com/linecorp/decaton/processor/CoreFunctionalityTest.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    @ClassRule
+    public static KafkaClusterRule rule = new KafkaClusterRule();
+
+    @Test(timeout = 30000)
+    public void testProcessConcurrent() {
+        ProcessorTestSuite
+                .builder(rule)
+                .configureProcessorsBuilder(builder -> builder.thenProcess((ctx, task) -> {
+                    // adding some pseudo random delay to generate out-of-order completion
+                    Thread.sleep(Math.abs(task.hashCode()) % 10);
```
[the remark this answers is not in the document]
Comment (ocadaruma) [reply processor/src/it/java/com/linecorp/decaton/processor/CoreFunctionalityTest.java:41]: My intention of `Math.abs(task.hashCode()) % 10` was to make test result stable and reproducible.

But, I'm thinking now it's better to introduce real randomness here (and also for other integration tests), by following reasons:
- This integration test tries to assert that semantics are always preserved, regardless of processing details
  * Then, processing content should be unpredictable as much as possible, to discover unknown issue in realistic environment
- If we found new case, (reproducible) unit test for it should be added

By the way I think there is one more perspective of integration test, that is "checking overall system's behavior (which cannot be tested individually) in reproducible environment". `MetricsTest` is an example of such test.

File: testing/src/main/java/com/linecorp/decaton/testing/processor/ReprocessAwareOrdering.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (ocadaruma) [reply testing/src/main/java/com/linecorp/decaton/testing/processor/ReprocessAwareOrdering.java:32]: Hmm but I agree with this point
https://github.com/line/decaton/pull/47#discussion_r436447109
> On the other hand it sounds bit wired to me that this framework has a built-in check for at-most-once because it's a "feature" outside of core functionality

Then it would be fine to name this as just `ProcessOrdering `.

File: processor/src/it/java/com/linecorp/decaton/processor/CoreFunctionalityTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (kawamuray) [reply processor/src/it/java/com/linecorp/decaton/processor/CoreFunctionalityTest.java:70]: ya, that's what I'm expecting.

File: processor/src/it/java/com/linecorp/decaton/processor/CoreFunctionalityTest.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+import com.linecorp.decaton.testing.KafkaClusterRule;
+import com.linecorp.decaton.testing.processor.ProcessingGuarantee.GuaranteeType;
+import com.linecorp.decaton.testing.processor.ProcessorTestSuite;
+
+public class CoreFunctionalityTest {
+    @ClassRule
+    public static KafkaClusterRule rule = new KafkaClusterRule();
+
+    @Test(timeout = 30000)
+    public void testProcessConcurrent() {
```
Comment (kawamuray) [inline processor/src/it/java/com/linecorp/decaton/processor/CoreFunctionalityTest.java:35]: How about adding 2 more tests wher's each of them checks the same thing but with `ProcessorScope.PARTITION` and `ProcessorScope.THREAD`?

File: processor/src/it/java/com/linecorp/decaton/processor/CoreFunctionalityTest.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        ProcessorTestSuite
+                .builder(rule)
+                .configureProcessorsBuilder(builder -> builder.thenProcess((ctx, task) -> {
+                    // adding some random delay to simulate realistic usage
+                    Thread.sleep(ThreadLocalRandom.current().nextLong(10L));
```
Comment (kawamuray) [inline processor/src/it/java/com/linecorp/decaton/processor/CoreFunctionalityTest.java:40]: Even though we agreed upon to introduce real randomness for this test, I think it's better to leave least hint for reproducing the same test (of course not guaranteed to be exactly same though) in case it failed.
How about using System.currentTimeMillis() as a seed of random along with printing it in log so that when the test fails, we can try reproducing it with reusing that seed for same random generation?

File: testing/src/main/java/com/linecorp/decaton/testing/processor/ProcessorTestSuite.java [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+        try {
+            producer = TestUtils.producer(rule.bootstrapServers());
+            for (int i = 0; i < subscriptions.length; i++) {
+                subscriptions[i] = subscriptionConstructor.apply(i);
+            }
+            CompletableFuture<Map<Integer, Long>> produceFuture =
+                    produceTasks(producer, topic, record -> semantics.forEach(g -> g.onProduce(record)));
+
+            // perform rolling restart
+            rollingRestartLatch.await();
```
Comment (kawamuray) [inline testing/src/main/java/com/linecorp/decaton/testing/processor/ProcessorTestSuite.java:191]: Seems like this block can be much visible with some refactoring like extracting each step out into private methods.

File: processor/src/it/java/com/linecorp/decaton/processor/CoreFunctionalityTest.java (hunk 10 above)
Comment (ocadaruma) [reply processor/src/it/java/com/linecorp/decaton/processor/CoreFunctionalityTest.java:40]: Sounds good. I'll fix

File: testing/src/main/java/com/linecorp/decaton/testing/processor/ProcessorTestSuite.java [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+
+        try {
+            producer = TestUtils.producer(rule.bootstrapServers());
+            for (int i = 0; i < subscriptions.length; i++) {
+                subscriptions[i] = newSubscription(i, topic, Optional.of(rollingRestartLatch));
+            }
+            CompletableFuture<Map<Integer, Long>> produceFuture =
+                    produceTasks(producer, topic, record -> semantics.forEach(g -> g.onProduce(record)));
+
+            performRollingRestart(subscriptions, i -> newSubscription(i, topic, Optional.empty()));
```
Comment (kawamuray) [inline testing/src/main/java/com/linecorp/decaton/testing/processor/ProcessorTestSuite.java:168]: `rollingRestartLatch.await()` is missing here?

Comment (ocadaruma) [reply testing/src/main/java/com/linecorp/decaton/testing/processor/ProcessorTestSuite.java:168]: oops nice catch

**Comment by kawamuray** [comment]
LGTM 👍  great work!
